engine/src/main/java/org/hibernate/validator/engine/HibernateConstrainedType.java

```diff
@@ -0,0 +1,31 @@
+/*
+ * Hibernate Validator, declare and validate application constraints
+ *
+ * License: Apache License, Version 2.0
+ * See the license.txt file in the root directory or <http://www.apache.org/licenses/LICENSE-2.0>.
+ */
+package org.hibernate.validator.engine;
+
+import java.util.List;
+
+import org.hibernate.validator.internal.util.classhierarchy.Filter;
+
+/**
+ * An interface that represents a type/class of a bean that will be validated.
+ * Based on this type set of constraints will be determined, and applied to the
+ * validated object.
+ *
+ * @author Marko Bekhta
+ * @since 6.1
+ */
+public interface HibernateConstrainedType<T> {
+
+	/**
+	 * @return a class of an object that will be validated.
+	 */
+	Class<T> getActuallClass();
+
+	List<HibernateConstrainedType<? super T>> getHierarchy(Filter... filters);
```

Comment:
`Filter` is a type from an `internal` package, whereas this interface isn't. That's a bit inconsistent (and there's nightly build in place which should spot this), I reckon this type should go under `internal`, too?

Reply:
hmmm, I missed this part (about `Filter` being an internal class). 

>I reckon this type should go under internal, too?

that's a good question... And the answer I suppose would be driven by the answer to the next question. How do we want the users to pass the mapping-name for free form validation? I was thinking that we would actually ask them to pass this `HibernateConstrainedType` as one of the parameters. Something like

```java
<T> Set<ConstraintViolation<T>> validate(T object, HibernateConstrainedType<T> mapping, Class<?>... groups);
```
then we would need this class to be in the public API. But if we decide that we would just ask users for a string
```java
<T> Set<ConstraintViolation<T>> validate(T object, String mapping, Class<?>... groups);
```
And later inside the method just wrap it with our own impl ... we could then move it to internal package.

`HibernateConstrainedType` being public interface gives us interesting opportunities. For example using this hierarchy method from it we could, in theory, build a  constraints hierarchy for free form validation.